sapicore/engine/synapse/__init__.py

```diff
@@ -0,0 +1,333 @@
+        Whether transmission delay values from one presynaptic element are the same for all postsynaptic targets.
+        Provides a speed-control tradeoff (one matrix multiplication vs. N vector multiplications).
+
+    weight_init_method: Callable
+        Weight initialization method from `torch.nn.init`.
+
+    connections: Tensor
+        2D binary integer mask matrix indicating which source unit -> destination unit connections are enabled.
+        Loggable attribute.
+
+    weights: Tensor
+        2D float matrix storing a weight for each pair of source and destination ensemble units.
+        Loggable attribute.
+
+    output: Tensor
+        1D tensor storing the result of W*x matrix multiplication in the current simulation step.
+        Loggable attribute.
+
+    """
+
+    _config_props_: tuple[str] = ("weight_max", "weight_min", "delay_ms", "simple_delays")
+    _loggable_props_: tuple[str] = ("weights", "connections", "output")
+
+    # declare loggable instance attributes registered as torch buffers.
+    connections: tensor
+    weights: tensor
+    output: tensor
+
+    def __init__(
+        self,
+        src_ensemble: [Neuron] = None,
+        dst_ensemble: [Neuron] = None,
+        weight_max: float = 1000.0,
+        weight_min: float = -1000.0,
+        delay_ms: float = 0.0,
+        simple_delays: bool = True,
+        weight_init_method: Callable = xavier_uniform_,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+
+        # model-related common instance attributes.
+        self.src_ensemble = src_ensemble
+        self.dst_ensemble = dst_ensemble
+
+        try:
+            self.matrix_shape = (dst_ensemble.num_units, src_ensemble.num_units)
+
+        except AttributeError:
+            self.matrix_shape = (1, 1)
+
+        self.num_units = torch.prod(torch.tensor(self.matrix_shape), 0).item()
+
+        self.weight_max = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device) + weight_max
+        self.weight_min = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device) + weight_min
+        self.weight_init_method = weight_init_method  # xavier_uniform by default.
+
+        # whether transmission delays from one presynaptic element are shared among postsynaptic targets.
+        self.simple_delays = simple_delays
+
+        if self.simple_delays:
+            # initialize default transmission delay values as a tensor of size `src_ensemble.num_units`.
+            self.delay_ms = torch.zeros(self.matrix_shape[1], dtype=torch.float, device=self.device) + delay_ms
+
+            # container for data pulled from the queue in this simulation step (input as seen by `dst_ensemble`).
+            self.delayed_data = torch.zeros(self.matrix_shape[1], dtype=torch.float, device=self.device)
+
+            # maintains delayed data queues
+            self.delay_queue = [
+                deque(torch.zeros(delay.int(), device=self.device)) for delay in self.delay_ms.int() / self.dt
```

Comment:
I forgot to mention the first time, but it should be `(self.delay_ms / self.dt).int()`. Or round. The same issue in `heterogenize`.